Allow attaching additional data to schedule helper blocks

Klikini opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change addresses the feature request [Weekly Schedule Helper - Multi state option](https://community.home-assistant.io/t/weekly-schedule-helper-multi-state-option/458532). It allows adding data to each block in a schedule helper, which then appears as attribute(s) on the helper entity when that block is active.

The data can be a single, scalar value, such as a string or number, in which case it will appear as an attribute named `Data` on the entity:

```yaml
schedule:
  thermostat_schedule:
    name: "Thermostat schedule"
    thursday:
      - from: "17:00:00"
        to: "23:00:00"
        data: 71
    friday:
      - from: "17:00:00"
        to: "23:00:00"
        data: 70
    saturday:
      - from: "07:00:00"
        to: "10:00:00"
        data: 74
      - from: "16:00:00"
        to: "23:00:00"
        data: 72
```

![example of single data attribute](https://github.com/home-assistant/core/assets/1693960/59fb8c09-57e2-4472-9c64-a3bd96f207f7)

It can also be an object with multiple properties, in which case they will appear as separate attributes on the entity:

```yaml
schedule:
  light_schedule:
    name: "Light schedule"
    wednesday:
      - from: "17:00:00"
        to: "21:00:00"
        data:
          brightness: 100
          color_temp: 4000
    thursday:
      - from: "17:00:00"
        to: "23:00:00"
        data:
          brightness: 90
          color_temp: 3500
    friday:
      - from: "07:00:00"
        to: "10:00:00"
        data:
          brightness: 80
          color_temp: 3000
      - from: "16:00:00"
        to: "23:00:00"
        data:
          brightness: 60
          color_temp: 2500
```

![example of multiple data attributes](https://github.com/home-assistant/core/assets/1693960/254b0504-f97b-4818-a0dd-d232ee69954e)

These can then be used for automations, allowing for much easier scheduling than creating a bunch of triggers:

![scheduling based on data](https://github.com/home-assistant/core/assets/1693960/0aeb3d83-cb89-46b8-ab0c-f3454963572d)

I was hoping to also update the frontend to display a YAML editor for the data field when clicking on schedule blocks instead of just a delete prompt, but I haven't had enough time to figure out how to do that yet. Regardless, I think this feature is still very useful in YAML, and that is how people in the feature request thread planned to use it anyway.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: https://github.com/home-assistant/home-assistant.io/pull/32560

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Klikini committed:
Add a new optional "data" key when defining time ranges for the schedule component that exposes the provided data in the state attributes of the schedule entity when that time range is active

Klikini committed:
Merge branch 'home-assistant:dev' into schedule-data

frenck commented on homeassistant/components/schedule/__init__.py:
```diff
@@ -109,16 +110,21 @@ def serialize_to_time(value: Any) -> Any:
     {
         vol.Required(CONF_FROM): vol.Coerce(str),
         vol.Required(CONF_TO): serialize_to_time,
+        vol.Optional(CONF_DATA): CUSTOM_DATA_SCHEMA,
```
- The schema can be inlined, no need to create a constant in this case. 
- Any reason this is limited/coerced to strings?

Klikini replied:
1. I created a constant because it's used twice, once in `TIME_RANGE_SCHEMA` and again in `STORAGE_TIME_RANGE_SCHEMA`. Should I inline it even though it will be repeated?
2. I coerced everything to strings because I wasn't sure how to represent "any scalar". Using `typing.Any` would theoretically allow nested dicts and lists (assuming the attribute code can handle that) but voluptuous didn't want to validate anything when I used it. Would `vol.Or(bool, str, int, float)` be more appropriate?

frenck replied:
1. if it was a huge schema used many times, I would agree; this is overkill imho.
2. your suggestion seems acceptable

Klikini replied:
I replaced the string coercion, so now it's
```py
CUSTOM_DATA_VALUE = vol.Or(bool, str, int, float)
CUSTOM_DATA_SCHEMA = vol.Or(vol.Schema({str: CUSTOM_DATA_VALUE}), CUSTOM_DATA_VALUE)
```

Expanding this would give:
```py
TIME_RANGE_SCHEMA: VolDictType = {
    vol.Required(CONF_FROM): cv.time,
    vol.Required(CONF_TO): deserialize_to_time,
    vol.Optional(CONF_DATA): vol.Or(vol.Schema({str: vol.Or(bool, str, int, float)}), vol.Or(bool, str, int, float)),
}

STORAGE_TIME_RANGE_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_FROM): vol.Coerce(str),
        vol.Required(CONF_TO): serialize_to_time,
        vol.Optional(CONF_DATA): vol.Or(vol.Schema({str: vol.Or(bool, str, int, float)}), vol.Or(bool, str, int, float)),
    }
)
```

This feels redundant and more fragile to maintain, as it's not clear that these have elements that need to match each other. If you'd still prefer this I can change it, but I think it's worth avoiding the repetition.

emontnemery replied:
I agree it's better to avoid repeating 👍

frenck commented on homeassistant/components/schedule/__init__.py:
```diff
@@ -344,6 +352,14 @@ def _update(self, _: datetime | None = None) -> None:
+
+        if isinstance(current_data, dict):
+            # Add each key/value pair in the data to the entity's state attributes
+            self._attr_extra_state_attributes.update(current_data)
```
This one is a little confusing IMHO, sometimes it is added to the attributes, or otherwise it is stored in a `data` key.

IMHO, this is magic we should avoid. Let's either always store it in a data key (or not).

Honestly, I think we should always use the data key, mostly because it would allow us to exclude it from the recorder (which is an change that should be added to this PR as well).

Klikini replied:
I like the idea of excluding a single key from the recorder history, but my goal with the "flattening" of data attributes into the entity's state attributes was to make it easier to build automations. How do you select, for example, `data.brightness` in an automation trigger?

![entity attributes](https://github.com/home-assistant/core/assets/1693960/a2b9cd6c-0bd4-4872-9c14-06ab6ca3b498)

![automation trigger](https://github.com/home-assistant/core/assets/1693960/515188cc-368d-4807-b59a-77bd84183008)

frenck replied:
Accessing those is harder indeed. Alternatively, we do flatten them, but exclude all attributes from being recorded in that case.

Klikini replied:
I just pushed a new commit that adds any custom attributes present to `_unrecorded_attributes`. Is that the preferred way to do this?

emontnemery replied:
That won't work, you'll need to also set the `Entity` class' private attribute `__combined_unrecorded_attributes`:
```
    self._unrecorded_attributes = frozenset(self.all_custom_data_keys())
    self._Entity__combined_unrecorded_attributes = (
            self._entity_component_unrecorded_attributes | self._unrecorded_attributes
        )
```

Please add a test which ensures the exclusion actually works, search for `test_exclude_attributes` and you'll find tests which do just that.

emontnemery replied:
The test update is not correct the state set is just:
```
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=schedule.test, old_state=None, new_state=<state schedule.test=off; editable=False, next_event=2024-08-03T09:00:00-07:00, icon=mdi:party-popper, friendly_name=Party mode @ 2024-08-02T08:51:58.895820-07:00>>
```

The test should assert the state has `CONF_DATA` and `"party_level"` state attributes, and then check those state attributes are not recorded.

Klikini replied:
I think this approach is valid (or at least, it also appears at https://github.com/home-assistant/core/blob/dev/tests/components/select/test_recorder.py#L49).

I just pushed a new commit that updates the Entity class's unrecorded attributes like you suggested and adds assertions that the attributes are present on the entity to the test (in addition to the existing assertions that it is not present in the recorded states).

Klikini replied:
I've tried a few different ways of mocking the current time that I found in other tests and none of them seem to work.

I've tried this:
```python
async_fire_time_changed(hass, datetime(2024, 8, 2, 7, 30))
await hass.async_block_till_done()
```

And this:
```python
with freeze_time(datetime(2024, 8, 2, 7, 30)):
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
```

(both with and without `tzinfo=dt_util.UTC`)

But following either of those methods with this:

```python
state = hass.states.get("schedule.test")
print(state)
```

...results in this output, which says Party mode @ the wrong time:

```
<state schedule.test=off; editable=False, next_event=2024-08-07T03:00:00-07:00, icon=mdi:party-popper, friendly_name=Party mode @ 2024-08-06T23:22:40.040380-07:00>
```

Is there any documentation on mocking times like this? I couldn't find any, and there seems to be no consensus among the other time-based tests that I can make sense of. Thank you for any help you can offer :)

emontnemery replied:
Passing a datetime object to `async_fire_time_changed` is the old way to implement time travelling in tests, it is very hard to get right because `async_fire_time_changed` only patches a few select time related functions.

Using freezegun is much preferred because it is much less likely to work since freezegun searches all modules for imported time functions and patched them.

When you use freezegun, you need to distinguish between the time zone on the machine running pytest and the timezone configured in the fake Home Assistant instance configured when running tests. The former is not predictable, the latter is `-07:00` unless the test is configured otherwise.

If you don't tell freezegun what timezone to use, it will assume the time zone used by the machine running pytest, which is not what you want.

The next problem is that the `schedule` enity sets up a listener for a future time, which won't fire if you travel backwards in time.

All in all, something like this should work:
```diff
diff --git a/tests/components/schedule/test_recorder.py b/tests/components/schedule/test_recorder.py
index e0ae47e19dd..6e752aad2ad 100644
--- a/tests/components/schedule/test_recorder.py
+++ b/tests/components/schedule/test_recorder.py
@@ -4,6 +4,7 @@ from __future__ import annotations

 from datetime import timedelta

+from freezegun.api import FrozenDateTimeFactory
 import pytest

 from homeassistant.components.recorder.history import get_significant_states
@@ -18,8 +19,9 @@ from tests.components.recorder.common import async_wait_recording_done


 @pytest.mark.usefixtures("recorder_mock", "enable_custom_integrations")
-async def test_exclude_attributes(hass: HomeAssistant) -> None:
+async def test_exclude_attributes(hass: HomeAssistant, freezer: FrozenDateTimeFactory) -> None:
     """Test attributes to be excluded."""
+    freezer.move_to("2024-08-02 06:30:00-07:00")  # Before Friday event
     now = dt_util.utcnow()
     assert await async_setup_component(
         hass,
@@ -50,8 +52,23 @@ async def test_exclude_attributes(hass: HomeAssistant) -> None:
     assert state.attributes[ATTR_ICON]
     assert state.attributes[ATTR_NEXT_EVENT]

+    # Move to during Friday event
+    freezer.move_to("2024-08-02 07:30:00-07:00")
+    async_fire_time_changed(hass, fire_all=True)
     await hass.async_block_till_done()
-    async_fire_time_changed(hass, dt_util.utcnow() + timedelta(minutes=5))
+    state = hass.states.get("schedule.test")
+    assert state.attributes["party_level"]
+
+    # Move to during Sunday event
+    freezer.move_to("2024-08-04 11:30:00-07:00")
+    async_fire_time_changed(hass, fire_all=True)
+    await hass.async_block_till_done()
+    state = hass.states.get("schedule.test")
+    assert state.attributes[CONF_DATA]
+
+    await hass.async_block_till_done()
+    freezer.tick(timedelta(minutes=5))
+    async_fire_time_changed(hass)
     await hass.async_block_till_done()
     await async_wait_recording_done(hass)
```

Klikini replied:
Ahh, that explains it! I tried specifying UTC because I thought the UTC-7 was coming from my computer's time zone, so I didn't realize that was _supposed_ to be the time zone used for testing.

I just pushed that change (with some tweaks to the asserts to make them more specific). Thank you again for the help!

frenck reviewed (dismissed):
Thanks for the PR @Klikini 👍 

- Please add tests
- Let's exclude this data from being recorded, as it is static data

../Frenck

Klikini committed:
Merge branch 'home-assistant:dev' into schedule-data

Klikini committed:
Merge type annotation additions from upstream

Klikini committed:
Exclude all schedule entry custom data attributes from the recorder (with tests)

Klikini committed:
Merge branch 'home-assistant:dev' into schedule-data

Klikini committed:
Fix setting schedule attributes to exclude from recorder, update test to verify the attributes exist but are not recorded

Klikini committed:
Fix test to ensure schedule data attributes are not recorded

emontnemery commented:
@Klikini Please don't forget to click the "Ready for review" button if you want the PR to be reviewed again.

emontnemery commented on homeassistant/components/schedule/__init__.py:
```diff
@@ -109,16 +110,22 @@ def serialize_to_time(value: Any) -> Any:
 }
 
+# Extra data that the user can set on each time range
+CUSTOM_DATA_VALUE = vol.Or(bool, str, int, float)
+CUSTOM_DATA_SCHEMA = vol.Or(vol.Schema({str: CUSTOM_DATA_VALUE}), CUSTOM_DATA_VALUE)
```
Let's use `vol.Any`, that's almost exclusively used throughout the codebase
```suggestion
CUSTOM_DATA_VALUE = vol.Any(bool, str, int, float)
CUSTOM_DATA_SCHEMA = vol.Any(vol.Schema({str: CUSTOM_DATA_VALUE}), CUSTOM_DATA_VALUE)
```

emontnemery reviewed (changes requested):
With the additional data now flattened, I think it's just confusing to allow either a single value which goes to `attributes["data"]` or a dictionary which is merged with `attributes`.

I think we should instead only support a dictionary with additional data.

Klikini committed:
Use vol.Any in place of vol.Or

Co-authored-by: Erik Montnemery <erik@montnemery.com>

emontnemery reviewed (commented):
Suggestion to remove the value-only data option

Klikini commented:
@emontnemery lol I was just finishing up my own changes to remove the shorthand data syntax locally, so I will reconcile our changes tomorrow and get a new commit pushed.

emontnemery commented:
> @emontnemery lol I was just finishing up my own changes to remove the shorthand data syntax locally, so I will reconcile our changes tomorrow and get a new commit pushed.

You can just ignore my suggested changes and push your own if you already have something which works

Klikini committed:
Remove schedule block custom data shorthand
as requested in https://github.com/home-assistant/core/pull/116585#pullrequestreview-2280260436

emontnemery commented on homeassistant/components/schedule/__init__.py:
```diff
@@ -253,6 +258,12 @@ def __init__(self, config: ConfigType, editable: bool) -> None:
         self._attr_name = self._config[CONF_NAME]
         self._attr_unique_id = self._config[CONF_ID]
 
+        # Exclude any custom attributes that may be present on time ranges.
```
```suggestion
        # Exclude any custom attributes that may be present on time ranges from recording.
```

emontnemery reviewed (approved):
LGTM, thanks @Klikini 👍

emontnemery committed:
Update homeassistant/components/schedule/__init__.py
